presto-spark/src/main/java/com/facebook/presto/spark/execution/SparkOutputBuffer.java

```diff
@@ -0,0 +1,191 @@
+                state.canAddPages(),
+                0,
+                0,
+                0,
+                0,
+                ImmutableList.of());
+    }
+
+    @Override
+    public ListenableFuture<?> isFull()
+    {
+        return NON_BLOCKED;
+    }
+
+    @Override
+    public void enqueue(Lifespan lifespan, List<SerializedPage> pages)
+    {
+        enqueue(lifespan, 0, pages);
+    }
+
+    @Override
+    public void enqueue(Lifespan lifespan, int partition, List<SerializedPage> pages)
+    {
+        checkState(state.get() == NO_MORE_BUFFERS, "unexpected state: %s", state.get());
+        pages.forEach(page -> bufferedPages.add(new PageWithPartition(page, partition)));
+    }
+
+    @Override
+    public void setNoMorePages()
+    {
+        state.setIf(FLUSHING, state -> state == NO_MORE_BUFFERS);
+    }
+
+    public boolean hasPagesBuffered()
+    {
+        return !bufferedPages.isEmpty();
+    }
+
+    public PageWithPartition getNext()
+    {
+        PageWithPartition page = bufferedPages.poll();
+        if (bufferedPages.isEmpty()) {
+            state.setIf(FINISHED, state -> state == FLUSHING);
+        }
+        return page;
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return state.get() == FINISHED;
+    }
+
+    @Override
+    public ListenableFuture<BufferResult> get(OutputBufferId bufferId, long token, DataSize maxSize)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void addStateChangeListener(StateChangeListener<BufferState> stateChangeListener)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void setOutputBuffers(OutputBuffers newOutputBuffers)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void acknowledge(OutputBufferId bufferId, long token)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void abort(OutputBufferId bufferId)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void destroy()
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void fail()
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void setNoMorePagesForLifespan(Lifespan lifespan)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void registerLifespanCompletionCallback(Consumer<Lifespan> callback)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public boolean isFinishedForLifespan(Lifespan lifespan)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public long getPeakMemoryUsage()
+    {
+        return 0;
+    }
+
+    @Override
+    public double getUtilization()
+    {
+        return 0;
+    }
+
+    @Override
+    public boolean isOverutilized()
```

Comment:
Just realized this `SparkOutputBuffer` never pushes back right? -- Will this cause excessive memory usage on Spark Task?

Reply:
Added a TODO item